python/lsst/meas/algorithms/loadReferenceObjects.py

```diff
@@ -290,146 +292,157 @@ def _makeBoxRegion(BBox, wcs, BBoxPadding):
         bbox : `lsst.geom.box2I`
-            Box which bounds a region in pixel space
+            Box which bounds a region in pixel space.
         wcs : `lsst.afw.geom.SkyWcs`
-            Wcs object defining the pixel to sky (and inverse) transform for the space
-            of pixels of the supplied bbox
-        filterName : `str`
-            Name of camera filter, or None or blank for the default filter
-        epoch : `astropy.time.Time` (optional)
-            Epoch to which to correct proper motion and parallax,
-            or None to not apply such corrections.
-        photoCalib : None
-            Deprecated and ignored, only included for api compatibility
-        bboxPadding : `int`
-            Number describing how much to pad the input bbox by (in pixels), defaults
-            to 100. This parameter is necessary because optical distortions in telescopes
-            can cause a rectangular pixel grid to map into a non "rectangular" spherical
-            region in sky coordinates. This padding is used to create a spherical
-            "rectangle", which will for sure enclose the input box.  This padding is only
-            used to determine if the reference catalog for a sky patch will be loaded from
-            the data store, this function will filter out objects which lie within the
-            padded region but fall outside the input bounding box region.
+            Wcs object defining the pixel to sky (and inverse) transform for
+            the supplied ``bbox``.
+        filterName : `str` or `None`, optional
+            Name of camera filter, or `None` or blank for the default filter.
+        epoch : `astropy.time.Time` or `None`, optional
+            Epoch to which to correct proper motion and parallax, or `None`
+            to not apply such corrections.
+        photoCalib : `None`
+            Deprecated and ignored, only included for api compatibility.
 
         Returns
         -------
         referenceCatalog : `lsst.afw.table.SimpleCatalog`
-            Catalog containing reference objects inside the specified bounding box
+            Catalog containing reference objects inside the specified bounding
+            box (padded by self.config.pixelMargin).
 
         Raises
         ------
-        `lsst.pex.exception.RuntimeError`
-            Raised if no reference catalogs could be found for the specified region
-
-        `lsst.pex.exception.TypeError`
-            Raised if the loaded reference catalogs do not have matching schemas
+        RuntimeError
+            Raised if no reference catalogs could be found for the specified
+            region.
+        TypeError
+            Raised if the loaded reference catalogs do not have matching
+            schemas.
         """
+        bboxPadding = self.config.pixelMargin
         innerSkyRegion, outerSkyRegion, _, _ = self._makeBoxRegion(bbox, wcs, bboxPadding)
+        paddedBbox = copy.deepcopy(bbox)
+        paddedBbox.grow(bboxPadding)
```

Comment:
I can believe this works in this particular case, but it is not solving the original problem that the boxPadding was intended for.

If you take an arbitrary bounding box and attempt to use the corners of the bounding box to define a convex polynomial on the sky, you can end up in a situation where there is area of the bounding box that does not actually fall inside the outer convex polynomial (since you are trying to put a rectangular box onto a spherical grid). You could use the same value for both, but it is an additional growth factor (see below for more)

The bboxPadding argument was designed to handle that issue and not growing the bbox to worry about astrometric errors. I designed this such that if you wanted a larger box, you would grow the box before calling this method.

However, I take it that the calling code does not do this per-say if we want it to be the same for gen2 and gen3 at this point.

The correct fix for this then would be to leave the bboxPadding argument (because it is independent of the pixelMargin parameter, and is not part of the gen2 implementation) and grow the bounding box here because it seems we must for what is going on with gen2.

Something like:
```
def loadPixelBox(self, bbox, wcs, filterName=None, epoch=None, photoCalib=None, bboxPadding=100):
    ...
    paddedBbox = Box2I(bbox)
    paddedBbox.grow(self.config.pixelMargin)
    innerSkyRegion, outerSkyRegion, _, _ = self._makeBoxRegion(paddedBbox, wcs, bboxPadding)
```

You could rename the `bboxPadding` parameter to reflect that it affects the generation of the spherical box creation instead, if you want to disambiguate.

What was written

Reply:
I see your point, but I still don't think this will work because the `region` getting passed to `loadRegion` will be the `outerSkyRegion` that only got the `bboxPadding` added to it (i.e. `config.pixelMargin` will again be effectively ignored).  It's also not possible for someone running a pipeline to have any control over the `bboxPadding` value (not to mention in the two places it gets a default of 100), which seems very dangerous as a fixed "number of pixels" default is unlikely to be appropriate for all cameras/detectors.

I would think that it's reasonable for the user to be responsible for configuring the `config.pixelMargin` to accommodate this case.  Also, I think with the way the refcats are partitioned and the fact each partition that has any amount of overlap with the `region` under consideration will get loaded, it should be really unlikely(unlucky!) that the effective loaded region (with proper `pixelMargin` padding) will leave us with any missing area.  If you really don't trust the `config.pixelMargin` setting, I think this could only work if you are taking `bboxPadding` as "in addition to", so that the `outerSkyRegion` gets padded with `config.pixelMargin + bboxPadding`...but that just becomes convoluted and confusing (but, even then, I'd recommend the default be set to 0, not 100).

So, I'm still advocating getting rid of this option.  Let me know what you think...if you disagree, then we still need a solution to the fact that, *at minimum*, the padding controlling the size of the load regions needs obey `config.pixelMargin`.   Perhaps we could add to the `pixelMargin` doc string to emphasize this point?

[I will work on explicitly testing my statement above with plots such as those I attached to the ticket, but that involves strategically placed breaks to persist the loaded catalog, and I have to remind myself where they go!  So please do consider the logic once again while I'm at it...and stop me if you think it's unnecessary overkill!]

Reply:
I do disagree, the `outerSkyRegion` must be effectively `config.pixelMargin + bboxPadding` which is what happened in the example I put in the comment (notice the argument change to `_makeBoxRegion`. You are correct that the `paddedBbox.grow()` is needed in this method because we cant yet change the logic in the calling code.

I dont think I explained my self adequately above. The problem is that if you have a bbox grown by `pixelMargin` from which you then generate a bounding convex polynomial without growing a second amount, you can get a situation where there is on sky area for which `bbox.contains(coord_converted_to_x_y)` reports `True` but `outerRegion.contains(coord)` reports `False` due to how the polygons get inscribed vs the tangent plane projection, which is what the bbox is.

This would lead to a situation where some ref_cats may get rejected because their stored region dose not intersect the outerRegion, but some records w/n that ref_cat would actually fall within the bounding box.

You are right that currently the code does give an option to change the call signature to pass an alternative padding, but I consider that a bug we need to resolve post gen2. This loader will need refactored anyway, and we really should be moving it futher away from being a subtask rather than trying to conform to the old way of doing it. I put that option in there, as it was large enough in all the testing I did, and I didn't want to yet add a new config option to parent task config classes, as then that would have to be supported well forward.

Reply:
Ah...sorry, I did miss that you passed in the `paddedBbox` as well as `bboxPadding` to `_makeBoxRegion`.  Ok -- while my concerns about the unchangeable default still stand -- I will make your suggested change  and do some further testing (as this should load more area than before, I think the filtered catalog should still end up looking the same...but an explicit check is warranted!)  I'll post on the ticket once I have some results.

What do you think of `bboxPaddingForSphere` as a name for this variable (that's terrible...please do suggest something better!!  But I am keen to be more specific in the name).

Reply:
The end catalog should be the same, in all but the weirdest cases (which we may not even ever hit with these instruments), but it would be one of those things that would be really hard for someone to trace down later if it did come up. 

maybe `bboxToSpherePadding` ? about as long as what you wrote. I'm not good at coming up with something either, which is how we got the confusing name to start with :smile:

Reply:
Ok, for now I've gone with the following (suggestions for improvement welcome):
```
        bboxToSpherePadding : `int`, optional                                                                  
            Padding to account for translating a set of corners into an                                        
            spherical (convex) boundary that is certain to encompase the                                       
            enitre area covered by the box.                             
```

Reply:
Wow...tips, dude!  We must have been typing simultaneously!